services/consensusalgo/benchmarkconsensus/leader.go

```diff
@@ -0,0 +1,198 @@
+	if err != nil {
+		panic(err)
+	}
+	return blockPair
+}
+
+func (s *service) leaderGenerateNewProposedBlockUnsafe() (*protocol.BlockPairContainer, error) {
+	s.reporting.Infof("Generating new proposed block for height %d", s.lastCommittedBlockHeight()+1)
+
+	// get tx
+	txOutput, err := s.consensusContext.RequestNewTransactionsBlock(&services.RequestNewTransactionsBlockInput{
+		BlockHeight: s.lastCommittedBlockHeight() + 1,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// get rx
+	rxOutput, err := s.consensusContext.RequestNewResultsBlock(&services.RequestNewResultsBlockInput{
+		BlockHeight: s.lastCommittedBlockHeight() + 1,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// generate signed block
+	if txOutput == nil || txOutput.TransactionsBlock == nil || rxOutput == nil || rxOutput.ResultsBlock == nil {
+		panic("invalid responses: missing fields")
+		// TODO: should we have these panics? because this is internal code
+	}
+	blockPair, err := s.leaderSignBlockProposal(txOutput.TransactionsBlock, rxOutput.ResultsBlock)
+	if err != nil {
+		return nil, err
+	}
+	return blockPair, nil
+}
+
+func (s *service) leaderSignBlockProposal(transactionsBlock *protocol.TransactionsBlockContainer, resultsBlock *protocol.ResultsBlockContainer) (*protocol.BlockPairContainer, error) {
+	blockPair := &protocol.BlockPairContainer{
+		TransactionsBlock: transactionsBlock,
+		ResultsBlock:      resultsBlock,
+	}
+
+	// prepare signature over the block headers
+	signedData := s.signedDataForBlockProof(blockPair)
+	sig, err := signature.SignEd25519(s.config.NodePrivateKey(), signedData)
+	if err != nil {
+		return nil, err
+	}
+
+	// generate block proof
+	blockPair.ResultsBlock.BlockProof = (&protocol.ResultsBlockProofBuilder{
+		Type: protocol.RESULTS_BLOCK_PROOF_TYPE_BENCHMARK_CONSENSUS,
+		BenchmarkConsensus: &consensus.BenchmarkConsensusBlockProofBuilder{
+			Sender: &consensus.BenchmarkConsensusSenderSignatureBuilder{
+				SenderPublicKey: s.config.NodePublicKey(),
+				Signature:       sig,
+			},
+		},
+	}).Build()
+	return blockPair, nil
+}
+
+func (s *service) leaderHandleCommittedVote(sender *gossipmessages.SenderSignature, status *gossipmessages.BenchmarkConsensusStatus) {
+	successfullyVotedBlock := blockHeightNone
+	defer func() {
+		// this needs to happen after s.mutex.Unlock() to avoid deadlock
+		if successfullyVotedBlock != blockHeightNone {
+			s.successfullyVotedBlocks <- successfullyVotedBlock
+		}
+	}()
+
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	// validate the vote
+	err := s.leaderValidateVoteUnsafe(sender, status)
+	if err != nil {
+		s.reporting.Error(err) // TODO: wrap with added context
+		return
+	}
+
+	// add the vote
+	s.lastCommittedBlockVoters[sender.SenderPublicKey().KeyForMap()] = true
+
+	// count if we have enough votes to move forward
+	existingVotes := len(s.lastCommittedBlockVoters) + 1
+	neededVotes := int(math.Ceil(float64(s.config.NetworkSize(0)) * 2 / 3))
```

Comment:
extract this calculation to a function and name according to meaning of the calc ("requiredQuorumSize"?)